tools/llvm-spirv/llvm-spirv.cpp

```diff
@@ -259,13 +278,110 @@ static int regularizeLLVM() {
   return 0;
 }
 
+static int parseSPVExtOption(
+    SPIRV::TranslatorOpts::ExtensionsStatusMap &ExtensionsStatus) {
+  // Map name -> id for known extensions
+  std::map<std::string, ExtensionID> ExtensionNamesMap;
+#define _STRINGIFY(X) #X
+#define STRINGIFY(X) _STRINGIFY(X)
+#define EXT(X) ExtensionNamesMap[STRINGIFY(X)] = ExtensionID::X;
+#include "LLVMSPIRVExtensions.inc"
+#undef EXT
+#undef STRINGIFY
+#undef _STRINGIFY
+
+  // Set the initial state
+  // During SPIR-V consumption, assume that any known extension is allowed
+  // During SPIR-V generation, assume that any known extension is disallowed
+  // During convertation to/from SPIR-V text representation, assume that any
+  // known extension is allowed
+  for (const auto &It : ExtensionNamesMap)
+    ExtensionsStatus[It.second] = IsReverse;
+
+  if (!SPVExt.getNumOccurrences())
+    return 0; // Nothing to do
+
+  size_t CurPos = 0;
+  std::set<std::string> EncounteredExtensions;
+  do {
+    // Check for plus or minus sign
+    if ('+' != SPVExt[CurPos] && '-' != SPVExt[CurPos]) {
+      errs() << "Invalid value of --spirv-ext, expected format is:\n"
+             << "\t--spirv-ext=+EXT_NAME,-EXT_NAME\n";
+      return -1;
+    }
+
+    // Find the next ',' to extract extension name
+    size_t CommaPos = SPVExt.find_first_of(',', CurPos);
+    // +1 to skip sign symbol
+    auto ExtName = SPVExt.substr(CurPos + 1, CommaPos - CurPos - 1);
+
+    // Reject empty strings
+    if (ExtName.empty()) {
+      errs() << "Invalid value of --spirv-ext, expected format is:\n"
+             << "\t--spirv-ext=+EXT_NAME,-EXT_NAME\n";
+      return -1;
+    }
+
+    bool ExtStatus = ('+' == SPVExt[CurPos]);
+    if ("all" == ExtName) {
+      // Update status for all known extensions
+      for (const auto &It : ExtensionNamesMap)
+        ExtensionsStatus[It.second] = ExtStatus;
+    } else {
+      // Reject unknown extensions
+      const auto &It = ExtensionNamesMap.find(ExtName);
+      if (ExtensionNamesMap.end() == It) {
+        errs() << "Unknown extension '" << ExtName << "' was specified via "
+               << "--spirv-ext option\n";
+        return -1;
+      }
+
+      ExtensionsStatus[It->second] = ExtStatus;
+    }
+
+    // Check for +ext,-ext pairs
```

Comment:
Should we have implementation aligned with -cl-ext ? If I recall correctly, for -cl-ext, if the same extension is encountered multiple times with possibly different signs the latter one takes precedence.

Reply:
Done